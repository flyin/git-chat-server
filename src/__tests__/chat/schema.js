const _ = require('lodash');
const { graphql } = require('graphql');

const { connection, mongoose } = require('../../services/mongoose');
const { Mutation } = require('../../resolvers/chat');
const { clearDatabase } = require('../../commands/clear-database');
const schema = require('../../schema');

describe('Chat schema test case', () => {
  beforeAll(() => connection);
  afterAll(() => mongoose.disconnect());
  beforeEach(() => clearDatabase());

  it('Should create message', async () => {
    const query = `
      mutation message($text: String!, $channelId: String!) {
        createMessage(text: $text, channelId: $channelId) {
          text
          channel {
            name
          }
        }
      }
    `;

    const channel = await Mutation.createChannel(null, { name: 'some-channel' });
    const result = await graphql(schema, query, {}, {}, { channelId: channel._id, text: 'Some message' });
    expect(result.data).toMatchSnapshot();
  });

  it('Should return channel messages', async () => {
    const query = `
      query messages($channelId: ID!) {
        getMessages(channelId: $channelId) {
          channel {
            name
          }
          text
        }
      }
    `;

    const channel = await Mutation.createChannel(null, { name: 'some-channel' });
    const anotherChannel = await Mutation.createChannel(null, { name: 'some-another-channel' });

    const messages = [
      { channelId: channel._id, text: 'message-1' },
      { channelId: channel._id, text: 'message-2' },
      { channelId: channel._id, text: 'message-3' },
      { channelId: channel._id, text: 'message-4' }
    ];

    const anotherMessages = [
      { channelId: anotherChannel._id, text: 'message-10' },
      { channelId: anotherChannel._id, text: 'message-11' },
      { channelId: anotherChannel._id, text: 'message-12' }
    ];

    await Promise.all([
      ...messages,
      ...anotherMessages
    ].map(({ channelId, text }) => Mutation.createMessage(null, { channelId, text })));

    const result = await graphql(schema, query, {}, {}, { channelId: channel._id });
    expect(result.data.getMessages.length).toBe(messages.length);

    result.data.getMessages.forEach(message => {
      expect(_.findIndex(messages, { text: message.text })).toBeGreaterThanOrEqual(0);
      expect(message).toMatchSnapshot(`message-${message.text}`);
    });
  });

  it('Should create channel', async () => {
    const channelName = 'test-foo';

    const query = `
      mutation channel($name: String!) {
        createChannel(name: $name){
          name
        }
      }
    `;

    const result = await graphql(schema, query, {}, {}, { name: channelName });
    expect(result.data).toMatchSnapshot();
    expect(result.data.createChannel.name).toBe(channelName);
  });

  it('Should return channel by id', async () => {
    const channelName = 'test-foo';
    const channel = await Mutation.createChannel(null, { name: channelName });

    const query = `
      query getChannel($channelId: ID!) {
        getChannel(channelId: $channelId) {
          name
        }
      }
    `;

    const result = await graphql(schema, query, {}, {}, { channelId: channel._id });
    expect(result.data).toMatchSnapshot();
    expect(result.data.getChannel.name).toBe(channelName);
  });

  it('Should return all channels', async () => {
    const channels = [
      { name: 'test-foo' },
      { name: 'test-foo1' },
      { name: 'asasd' }
    ];

    await Promise.all(channels.map(({ name }) => Mutation.createChannel(null, { name })));

    const query = `
      {
        getChannels {
          name
        }
      }
    `;

    const result = await graphql(schema, query);
    expect(result.data).toMatchSnapshot();
    expect(result.data.getChannels.length).toBe(channels.length);
  });
});
