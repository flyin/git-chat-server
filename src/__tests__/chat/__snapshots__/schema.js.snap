// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Chat schema test case Should create channel 1`] = `
Object {
  "createChannel": Object {
    "name": "test-foo",
  },
}
`;

exports[`Chat schema test case Should create message 1`] = `
Object {
  "createMessage": Object {
    "channel": Object {
      "name": "some-channel",
    },
    "text": "Some message",
  },
}
`;

exports[`Chat schema test case Should return all channels 1`] = `
Object {
  "getChannels": Array [
    Object {
      "name": "asasd",
    },
    Object {
      "name": "test-foo",
    },
    Object {
      "name": "test-foo1",
    },
  ],
}
`;

exports[`Chat schema test case Should return channel by id 1`] = `
Object {
  "getChannel": Object {
    "name": "test-foo",
  },
}
`;

exports[`message-message-1 1`] = `
Object {
  "channel": Object {
    "name": "some-channel",
  },
  "text": "message-1",
}
`;

exports[`message-message-2 1`] = `
Object {
  "channel": Object {
    "name": "some-channel",
  },
  "text": "message-2",
}
`;

exports[`message-message-3 1`] = `
Object {
  "channel": Object {
    "name": "some-channel",
  },
  "text": "message-3",
}
`;

exports[`message-message-4 1`] = `
Object {
  "channel": Object {
    "name": "some-channel",
  },
  "text": "message-4",
}
`;
